### PR TITLE
KTOR-6807 Add version variables

### PR DESCRIPTION
**Pull request by bjhham**
This introduces support for version variables, to improve the organization of the templated project.

Now when authoring a plugin, you can reference the versions with $ and declare them in the same yaml file (or from other version yaml files).

For example:
```yaml
3.0.+: org.apache.commons:commons-compress:$apacheCompress

apacheCompress: 1.26.2
```

This will result in the introduction of this line in `gradle.properties`:
```
apache_compress_version: 1.26.2
```

and this line in `build.gradle.kts`
```
val apache_compress_version by project
```

I've updated several of the plugins to use versions this way, so we'll have less of a mix of inline vs. property versions.  For future work, I'd like to introduce version catalogs as an option in the project generator, since it's more standard now with gradle projects and it will work in amper as well.

Bonus fix: the mongo plugin was incompatible with 3.0+ ktor due to the environment.monitor reference, so I updated it.

### CODE REVIEW

File: buildSrc/src/main/kotlin/io/ktor/plugins/registry/PluginReference.kt [hunk 1]
```diff
@@ -22,7 +22,7 @@ import java.io.File
  * @property versions A map of version ranges to artifacts for the plugin.
  * @property client A flag indicating whether the plugin is a client plugin.
  */
-@Serializable
+// @Serializable
```
Comment (e5l) [inline buildSrc/src/main/kotlin/io/ktor/plugins/registry/PluginReference.kt:25]: Please check if this still needed

File: gradle/libs.versions.toml [hunk 2]
```diff
@@ -1,5 +1,5 @@
 [versions]
-kotlin-version = "1.9.24"
+kotlin-version = "1.9.23"
```
Comment (e5l) [inline gradle/libs.versions.toml:2]: Could you tell me if we can start using 2.0.0?

File: src/test/resources/server/com.fail/bad_semver/versions.ktor.yaml [hunk 3]
```diff
@@ -1 +1 @@
-!!!: what??
\ No newline at end of file
+$$$$: what??
```
Comment (e5l) [inline src/test/resources/server/com.fail/bad_semver/versions.ktor.yaml:1]: 😁

**Review by e5l — APPROVED** [review]

File: buildSrc/src/main/kotlin/io/ktor/plugins/registry/PluginReference.kt (hunk 1 above)
Comment (bjhham) [reply buildSrc/src/main/kotlin/io/ktor/plugins/registry/PluginReference.kt:25]: Yeah it's no longer used, I'll remove.

File: gradle/libs.versions.toml (hunk 2 above)
Comment (bjhham) [reply gradle/libs.versions.toml:2]: Seems to be a non-trivial change on account of changes to the compiler API.  I'll look into it in a separate task.

**Pull request merged**